### ISSUE TITLE
sam_model_fast_registry ， Speed problem

### ISSUE THREAD

**Issue by KnSun99** [pytorch-labs/segment-anything-fast#80]
Hello,

I used segment-anything-fast with no significant acceleration compared to segment-anything.
Is the method incorrect?

code:
import cv2
import numpy as np
from segment_anything_fast import sam_model_registry, sam_model_fast_registry, SamPredictor

video_path = 'video.avi'
sam_checkpoint = 'sam_vit_h_4b8939.pth'

input_point = np.array([[305, 235]])
input_label = np.array([1])

video_cap = cv2.VideoCapture(video_path)
ret, image = video_cap.read()
sam = sam_model_fast_registry["vit_h"](checkpoint=sam_checkpoint)
sam.to(device="cuda:0")
predictor = SamPredictor(sam)
predictor.set_image(image)
masks, scores, logits = predictor.predict(point_coords=input_point, point_labels=input_label, multimask_output=True)
mask_input = logits[np.argmax(scores), :, :]

i_fn = 0
while video_cap.isOpened() and i_fn < nframes:
    ret, image = video_cap.read()
    if ret == True:
        predictor.set_image(image)
        masks, _, _ = predictor.predict(point_coords=input_point, point_labels=input_label, mask_input=mask_input[None, :, :], multimask_output=False)
    i_fn += 1

How do I set up batch input image? (set batch size)

Use onnxruntime?

Thanks

**Comment by msaroufim** [comment]
Could you please sahre more details on your setup including which PyTorch version you're using and which GPU type you're using?

**Comment by KnSun99** [comment]
![857be6d59e3eb680c15f14b9ae8dc0a](https://github.com/pytorch-labs/segment-anything-fast/assets/123467428/add2ccd4-3a7a-41ee-a827-5a7228ac16bb)


channels:
  - conda-forge
  - defaults
dependencies:
  - _libgcc_mutex=0.1=main
  - _openmp_mutex=5.1=1_gnu
  - alabaster=0.7.13=pyhd8ed1ab_0
  - arrow=1.3.0=pyhd8ed1ab_0
  - astroid=3.0.1=py311h38be061_0
  - asttokens=2.4.1=pyhd8ed1ab_0
  - atomicwrites=1.4.1=pyhd8ed1ab_0
  - attrs=23.1.0=pyh71513ae_1
  - autopep8=2.0.4=pyhd8ed1ab_0
  - babel=2.13.1=pyhd8ed1ab_0
  - beautifulsoup4=4.12.2=pyha770c72_0
  - binaryornot=0.4.4=py_1
  - black=23.10.1=py311h38be061_0
  - bleach=6.1.0=pyhd8ed1ab_0
  - brotli-python=1.0.9=py311h6a678d5_7
  - bzip2=1.0.8=h7b6447c_0
  - ca-certificates=2023.11.17=hbcca054_0
  - certifi=2023.11.17=pyhd8ed1ab_0
  - cffi=1.16.0=py311h5eee18b_0
  - chardet=5.2.0=py311h38be061_1
  - charset-normalizer=3.3.2=pyhd8ed1ab_0
  - click=8.1.7=unix_pyh707e725_0
  - cloudpickle=3.0.0=pyhd8ed1ab_0
  - colorama=0.4.6=pyhd8ed1ab_0
  - comm=0.1.4=pyhd8ed1ab_0
  - cookiecutter=2.5.0=pyhca7485f_0
  - cryptography=41.0.3=py311hdda0065_0
  - dbus=1.13.18=hb2f20db_0
  - debugpy=1.6.7=py311h6a678d5_0
  - decorator=5.1.1=pyhd8ed1ab_0
  - defusedxml=0.7.1=pyhd8ed1ab_0
  - diff-match-patch=20230430=pyhd8ed1ab_0
  - dill=0.3.7=pyhd8ed1ab_0
  - docstring-to-markdown=0.13=pyhd8ed1ab_0
  - docutils=0.20.1=py311h38be061_2
  - entrypoints=0.4=pyhd8ed1ab_0
  - exceptiongroup=1.2.0=pyhd8ed1ab_0
  - executing=2.0.1=pyhd8ed1ab_0
  - expat=2.2.10=h9c3ff4c_0
  - flake8=6.1.0=pyhd8ed1ab_0
  - fontconfig=2.14.1=hef1e5e3_0
  - freetype=2.10.4=h0708190_1
  - glib=2.69.1=he621ea3_2
  - gst-plugins-base=1.14.1=h6a678d5_1
  - gstreamer=1.14.1=h5eee18b_1
  - icu=58.2=hf484d3e_1000
  - idna=3.4=pyhd8ed1ab_0
  - imagesize=1.4.1=pyhd8ed1ab_0
  - importlib-metadata=6.8.0=pyha770c72_0
  - importlib_metadata=6.8.0=hd8ed1ab_0
  - importlib_resources=6.1.1=pyhd8ed1ab_0
  - inflection=0.5.1=pyh9f0ad1d_0
  - intervaltree=3.1.0=pyhd8ed1ab_1
  - ipykernel=6.26.0=pyhf8b6a83_0
  - ipython=8.17.2=pyh41d4057_0
  - isort=5.12.0=pyhd8ed1ab_1
  - jaraco.classes=3.3.0=pyhd8ed1ab_0
  - jedi=0.19.1=pyhd8ed1ab_0
  - jeepney=0.8.0=pyhd8ed1ab_0
  - jellyfish=1.0.1=py311hb02cf49_0
  - jinja2=3.1.2=pyhd8ed1ab_1
  - jpeg=9e=h166bdaf_1
  - jsonschema=4.20.0=pyhd8ed1ab_0
  - jsonschema-specifications=2023.11.1=pyhd8ed1ab_0
  - jupyter_client=8.6.0=pyhd8ed1ab_0
  - jupyter_core=5.5.0=py311h38be061_0
  - jupyterlab_pygments=0.2.2=pyhd8ed1ab_0
  - keyring=24.3.0=py311h38be061_0
  - krb5=1.20.1=h143b758_1
  - ld_impl_linux-64=2.38=h1181459_1
  - libclang=10.0.1=default_hb85057a_2
  - libedit=3.1.20221030=h5eee18b_0
  - libevent=2.1.12=hdbd6064_1
  - libffi=3.4.4=h6a678d5_0
  - libgcc-ng=11.2.0=h1234567_1
  - libgomp=11.2.0=h1234567_1
  - libllvm10=10.0.1=he513fc3_3
  - libpng=1.6.39=h5eee18b_0
  - libpq=12.15=hdbd6064_1
  - libsodium=1.0.18=h36c2ea0_1
  - libspatialindex=1.9.3=h9c3ff4c_4
  - libstdcxx-ng=11.2.0=h1234567_1
  - libuuid=1.41.5=h5eee18b_0
  - libxcb=1.15=h7f8727e_0
  - libxkbcommon=1.0.1=hfa300c1_0
  - libxml2=2.9.14=h74e7548_0
  - lz4-c=1.9.4=h6a678d5_0
  - markdown-it-py=3.0.0=pyhd8ed1ab_0
  - markupsafe=2.1.1=py311h5eee18b_0
  - matplotlib-inline=0.1.6=pyhd8ed1ab_0
  - mccabe=0.7.0=pyhd8ed1ab_0
  - mdurl=0.1.0=pyhd8ed1ab_0
  - mistune=3.0.2=pyhd8ed1ab_0
  - more-itertools=10.1.0=pyhd8ed1ab_0
  - mypy_extensions=1.0.0=pyha770c72_0
  - nbclient=0.8.0=pyhd8ed1ab_0
  - nbconvert=7.11.0=pyhd8ed1ab_0
  - nbconvert-core=7.11.0=pyhd8ed1ab_0
  - nbconvert-pandoc=7.11.0=pyhd8ed1ab_0
  - nbformat=5.9.2=pyhd8ed1ab_0
  - ncurses=6.4=h6a678d5_0
  - nest-asyncio=1.5.8=pyhd8ed1ab_0
  - nspr=4.35=h6a678d5_0
  - nss=3.89.1=h6a678d5_0
  - numpydoc=1.5.0=pyhd8ed1ab_0
  - openssl=3.0.12=h7f8727e_0
  - packaging=23.2=pyhd8ed1ab_0
  - pandoc=2.19.2=ha770c72_0
  - pandocfilters=1.5.0=pyhd8ed1ab_0
  - parso=0.8.3=pyhd8ed1ab_0
  - pathspec=0.11.2=pyhd8ed1ab_0
  - pcre=8.45=h9c3ff4c_0
  - pexpect=4.8.0=pyh1a96a4e_2
  - pickleshare=0.7.5=py_1003
  - pip=23.3.1=py311h06a4308_0
  - pkgutil-resolve-name=1.3.10=pyhd8ed1ab_1
  - platformdirs=4.0.0=pyhd8ed1ab_0
  - pluggy=1.3.0=pyhd8ed1ab_0
  - ply=3.11=py_1
  - prompt-toolkit=3.0.41=pyha770c72_0
  - prompt_toolkit=3.0.41=hd8ed1ab_0
  - psutil=5.9.0=py311h5eee18b_0
  - ptyprocess=0.7.0=pyhd3deb0d_0
  - pure_eval=0.2.2=pyhd8ed1ab_0
  - pycodestyle=2.11.1=pyhd8ed1ab_0
  - pycparser=2.21=pyhd8ed1ab_0
  - pydocstyle=6.3.0=pyhd8ed1ab_0
  - pyflakes=3.1.0=pyhd8ed1ab_0
  - pygments=2.17.2=pyhd8ed1ab_0
  - pylint=3.0.2=pyhd8ed1ab_0
  - pylint-venv=3.0.3=pyhd8ed1ab_0
  - pyls-spyder=0.4.0=pyhd8ed1ab_0
  - pyqt=5.15.10=py311h6a678d5_0
  - pyqt5-sip=12.13.0=py311h5eee18b_0
  - pyqtwebengine=5.15.10=py311h6a678d5_0
  - pysocks=1.7.1=pyha2e5f31_6
  - python=3.11.5=h955ad1f_0
  - python-dateutil=2.8.2=pyhd8ed1ab_0
  - python-fastjsonschema=2.19.0=pyhd8ed1ab_0
  - python-lsp-black=1.3.0=pyhd8ed1ab_0
  - python-lsp-jsonrpc=1.1.2=pyhd8ed1ab_0
  - python-lsp-server=1.9.0=pyhd8ed1ab_0
  - python-lsp-server-base=1.9.0=pyhd8ed1ab_0
  - python-slugify=8.0.1=pyhd8ed1ab_2
  - python_abi=3.11=2_cp311
  - pytoolconfig=1.2.5=pyhd8ed1ab_0
  - pytz=2023.3.post1=pyhd8ed1ab_0
  - pyxdg=0.28=pyhd8ed1ab_0
  - pyyaml=6.0.1=py311h5eee18b_0
  - pyzmq=25.1.0=py311h6a678d5_0
  - qdarkstyle=3.2=pyhd8ed1ab_0
  - qstylizer=0.2.2=pyhd8ed1ab_0
  - qt-main=5.15.2=h327a75a_7
  - qt-webengine=5.15.9=hd2b0992_4
  - qtawesome=1.2.3=pyhd8ed1ab_0
  - qtconsole=5.5.1=pyhd8ed1ab_0
  - qtconsole-base=5.5.1=pyha770c72_0
  - qtpy=2.4.1=pyhd8ed1ab_0
  - readline=8.2=h5eee18b_0
  - referencing=0.31.0=pyhd8ed1ab_0
  - requests=2.31.0=pyhd8ed1ab_0
  - rich=13.7.0=pyhd8ed1ab_0
  - rope=1.11.0=pyhd8ed1ab_1
  - rpds-py=0.10.6=py311hb02cf49_0
  - rtree=1.1.0=py311h3bb2b0f_0
  - secretstorage=3.3.3=py311h38be061_2
  - setuptools=68.0.0=py311h06a4308_0
  - sip=6.7.12=py311h6a678d5_0
  - six=1.16.0=pyh6c4a22f_0
  - snowballstemmer=2.2.0=pyhd8ed1ab_0
  - sortedcontainers=2.4.0=pyhd8ed1ab_0
  - soupsieve=2.5=pyhd8ed1ab_1
  - sphinx=7.2.6=pyhd8ed1ab_0
  - sphinxcontrib-applehelp=1.0.7=pyhd8ed1ab_0
  - sphinxcontrib-devhelp=1.0.5=pyhd8ed1ab_0
  - sphinxcontrib-htmlhelp=2.0.4=pyhd8ed1ab_0
  - sphinxcontrib-jsmath=1.0.1=pyhd8ed1ab_0
  - sphinxcontrib-qthelp=1.0.6=pyhd8ed1ab_0
  - sphinxcontrib-serializinghtml=1.1.9=pyhd8ed1ab_0
  - spyder=5.5.0=py311h38be061_1
  - spyder-kernels=2.5.0=unix_pyhd8ed1ab_0
  - sqlite=3.41.2=h5eee18b_0
  - stack_data=0.6.2=pyhd8ed1ab_0
  - text-unidecode=1.3=pyhd8ed1ab_1
  - textdistance=4.5.0=pyhd8ed1ab_0
  - three-merge=0.1.1=pyh9f0ad1d_0
  - tinycss2=1.2.1=pyhd8ed1ab_0
  - tk=8.6.12=h1ccaba5_0
  - tomli=2.0.1=pyhd8ed1ab_0
  - tomlkit=0.12.3=pyha770c72_0
  - tornado=6.3.3=py311h5eee18b_0
  - traitlets=5.13.0=pyhd8ed1ab_0
  - types-python-dateutil=2.8.19.14=pyhd8ed1ab_0
  - typing-extensions=4.8.0=hd8ed1ab_0
  - typing_extensions=4.8.0=pyha770c72_0
  - ujson=5.4.0=py311h6a678d5_0
  - urllib3=2.1.0=pyhd8ed1ab_0
  - watchdog=3.0.0=py311h38be061_1
  - wcwidth=0.2.12=pyhd8ed1ab_0
  - webencodings=0.5.1=pyhd8ed1ab_2
  - whatthepatch=1.0.5=pyhd8ed1ab_0
  - wheel=0.41.2=py311h06a4308_0
  - wurlitzer=3.0.3=pyhd8ed1ab_0
  - xz=5.4.2=h5eee18b_0
  - yaml=0.2.5=h7f98852_2
  - yapf=0.40.1=pyhd8ed1ab_0
  - zeromq=4.3.4=h2531618_0
  - zipp=3.17.0=pyhd8ed1ab_0
  - zlib=1.2.13=h5eee18b_0
  - zstd=1.5.5=hc292b87_0
  - pip:
      - contourpy==1.2.0
      - cycler==0.12.1
      - filelock==3.13.1
      - fonttools==4.45.1
      - fsspec==2023.10.0
      - h5py==3.10.0
      - kiwisolver==1.4.5
      - llvmlite==0.41.1
      - matplotlib==3.8.2
      - mpmath==1.3.0
      - networkx==3.2.1
      - numba==0.58.1
      - numpy==1.26.2
      - nvidia-cublas-cu12==12.1.3.1
      - nvidia-cuda-cupti-cu12==12.1.105
      - nvidia-cuda-nvrtc-cu12==12.1.105
      - nvidia-cuda-runtime-cu12==12.1.105
      - nvidia-cudnn-cu12==8.9.2.26
      - nvidia-cufft-cu12==11.0.2.54
      - nvidia-curand-cu12==10.3.2.106
      - nvidia-cusolver-cu12==11.4.5.107
      - nvidia-cusparse-cu12==12.1.0.106
      - nvidia-nccl-cu12==2.18.1
      - nvidia-nvjitlink-cu12==12.3.101
      - nvidia-nvtx-cu12==12.1.105
      - opencv-python==4.8.1.78
      - pandas==2.1.3
      - pillow==10.1.0
      - pyparsing==3.1.1
      - scipy==1.11.4
      - sympy==1.12
      - torch==2.1.1
      - torchaudio==2.1.1
      - torchvision==0.16.1
      - tqdm==4.66.1
      - triton==2.1.0
      - tzdata==2023.3

**Comment by msaroufim** [comment]
Yeah we need to imrpove autotuning heuristics for 4090, the blog showed perf numbers on A100

**Comment by KnSun99** [comment]
Thank you for your reply.
I look forward to the new version.

**Comment by cpuhrsch** [comment]
@KnSun99 - I just pushed a change to `sam_model_fast_registry` to pick a different default dtype. Also you'll need to install the latest pytorch nightly to run this code. Maybe those changes will help you make use of this on the 4090 today!

**Comment by KnSun99** [comment]
Thank you for your reply! I will try to run it.

**Comment by cpuhrsch** [comment]
@KnSun99 - Did this fix your issue?

**Comment by KnSun99** [comment]
Solved this problem.
Speed has increased.
Thank you.
